tac-operation-lifecycle/tac-operation-lifecycle-logic/src/database.rs

```diff
@@ -0,0 +1,1167 @@
+                    // [assume bathes are sorted by `start` field ascending at that point]
+                    let mut updated_watermark: Option<u64> = None;
+                    if let Some(last_interval_from_batch) = chunk.last() {
+                        updated_watermark =
+                            Some(last_interval_from_batch.end.clone().unwrap() as u64);
+                        let _ = self
+                            .set_watermark_internal(&tx, updated_watermark.unwrap())
+                            .await;
+                    }
+
+                    tx.commit().await?;
+                    tracing::debug!(
+                        "Successfully saved batch of {} intervals and updated watermark to {}",
+                        chunk.len(),
+                        if let Some(wm) = updated_watermark {
+                            wm.to_string()
+                        } else {
+                            "[NOT_UPDATED]".to_string()
+                        },
+                    );
+                }
+                Err(e) => {
+                    tx.rollback().await?;
+                    tracing::error!("Failed to save batch: {}", e);
+                    return Err(e.into());
+                }
+            }
+        }
+
+        Ok(intervals.len())
+    }
+
+    // stage_types is a vector Id->StageName
+    pub async fn register_stage_types(
+        &self,
+        stage_types: &HashMap<i32, String>,
+    ) -> anyhow::Result<()> {
+        for (stage_id, stage_name) in stage_types.into_iter() {
+            let stage_type_model = stage_type::ActiveModel {
+                id: Set(*stage_id),
+                name: Set(stage_name.to_string()),
+            };
+
+            stage_type::Entity::insert(stage_type_model)
+                .on_conflict(
+                    OnConflict::column(stage_type::Column::Id)
+                        .update_columns([stage_type::Column::Name])
+                        .to_owned(),
+                )
+                .exec(self.db.as_ref())
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    pub async fn insert_pending_operations(
+        &self,
+        operations: &ApiOperations,
+    ) -> anyhow::Result<()> {
+        let thread_id = thread::current().id();
+
+        // Start a transaction
+        let txn = self.db.begin().await?;
+
+        // Save all operations
+        for op in operations {
+            let operation_model = operation::ActiveModel {
+                id: Set(op.id.clone()),
+                operation_type: Set(None),
+                timestamp: Set(op.timestamp as i64),
+                status: Set(EntityStatus::Pending.to_id() as i32),
+                next_retry: Set(None),
+                retry_count: Set(0), // Initialize retry count
+            };
+
+            tracing::debug!(
+                "[Thread {:?}] Attempting to insert operation: {:?}",
+                thread_id,
+                operation_model
+            );
```

Comment:
```suggestion
            tracing::debug!(
                thread = thread_id,
                "Attempting to insert operation: {:?}",
                operation_model
            );
```
FYI you can do smth like this, or since it's repeated in other logs, maybe put it in a span?

Reply:
same for other logs

Reply:
I completely removed thread ID logging — it was just for optimization and doesn't seem needed anymore https://github.com/blockscout/blockscout-rs/pull/1325/commits/c024b40c2dc4f2da193b7e375f6ffc1a4c38633c